tower-balance/src/lib.rs

```diff
@@ -300,31 +345,103 @@ where
     /// When `Async::Ready` is returned, `chosen_ready_index` is set with a valid index
     /// into `ready` referring to a `Service` that is ready to disptach a request.
     fn poll_ready(&mut self) -> Poll<(), Self::Error> {
-        // Clear before `ready` is altered.
-        self.chosen_ready_index = None;
+        self.poll_ready_inner(D::Service::poll_ready)
+    }
 
-        // Before `ready` is altered, check the readiness of the last-used service, moving it
-        // to `not_ready` if appropriate.
-        if let Some(idx) = self.dispatched_ready_index.take() {
-            // XXX Should we handle per-endpoint errors?
-            self.poll_ready_index(idx).expect("invalid dispatched ready key")?;
-        }
+    fn call(&mut self, request: Request) -> Self::Future {
+        self.call(D::Service::call, request)
+    }
+}
 
-        // Update `not_ready` and `ready`.
-        self.update_from_discover()?;
-        self.promote_to_ready()?;
+impl<D, C, Request> DirectService<Request> for Balance<D, C>
+where
+    D: Discover,
+    D::Service: DirectService<Request>,
+    C: Choose<D::Key, D::Service>,
+{
+    type Response = <D::Service as DirectService<Request>>::Response;
+    type Error = Error<<D::Service as DirectService<Request>>::Error, D::Error>;
+    type Future = ResponseFuture<<D::Service as DirectService<Request>>::Future, D::Error>;
 
-        // Choose the next service to be used by `call`.
-        self.choose_and_poll_ready()
+    fn poll_ready(&mut self) -> Poll<(), Self::Error> {
+        self.poll_ready_inner(D::Service::poll_ready)
     }
 
     fn call(&mut self, request: Request) -> Self::Future {
-        let idx = self.chosen_ready_index.take().expect("not ready");
-        let (_, svc) = self.ready.get_index_mut(idx).expect("invalid chosen ready index");
-        self.dispatched_ready_index = Some(idx);
+        self.call(D::Service::call, request)
+    }
 
-        let rsp = svc.call(request);
-        ResponseFuture(rsp, PhantomData)
+    fn poll_service(&mut self) -> Result<Async<()>, Self::Error> {
+        let ready = &mut self.ready;
+        let not_ready = &mut self.not_ready;
+        let mut err = None;
+        self.pending.retain(|k| {
```

Comment:
I don't see where any keys are inserted into this set.

Reply:
Ah, yes, you're completely right. How silly of me. Sadly, I think we'll have to require `D::Key: Clone` to be able to maintain `pending` in a reasonable fashion. Maybe we should just get rid of the optimization?